interfaces/builtin/kubernetes_support.go

```diff
@@ -35,65 +47,226 @@ const kubernetesSupportBaseDeclarationSlots = `
-# is restricted because it gives wide ranging access to the host and other
-# processes.
+const kubernetesSupportConnectedPlugAppArmorCommon = `
+# Common rules for running as a kubernetes node
 
-# what is this for?
-#include <abstractions/dbus-strict>
-
-# Allow reading all information regarding cgroups for all processes
+# reading cgroups
 capability sys_resource,
-@{PROC}/@{pid}/cgroup r,
 /sys/fs/cgroup/{,**} r,
 
-# Allow adjusting the OOM score for Docker containers. Note, this allows
-# adjusting for all processes, not just containers.
+# Allow adjusting the OOM score for containers. Note, this allows adjusting for
+# all processes, not just containers.
 @{PROC}/@{pid}/oom_score_adj rw,
-/sys/kernel/mm/hugepages/ r,
+@{PROC}/sys/vm/overcommit_memory rw,
+/sys/kernel/mm/hugepages/{,**} r,
+
+capability dac_override,
+
+/usr/bin/systemd-run Cxr -> systemd_run,
+profile systemd_run (attach_disconnected,mediate_deleted) {
+  # Common rules for kubernetes use of systemd_run
+  #include <abstractions/base>
+
+  /usr/bin/systemd-run r,
+  owner @{PROC}/@{pid}/stat r,
+  owner @{PROC}/@{pid}/environ r,
+  @{PROC}/cmdline r,
+
+  # setsockopt()
+  capability net_admin,
+
+  /run/systemd/private rw,
+  /bin/true ixr,
+  ptrace trace peer=unconfined,
+###KUBERNETES_SUPPORT_SYSTEMD_RUN###
+}
+`
+
+const kubernetesSupportConnectedPlugAppArmorKubelet = `
+# Allow running as the kubelet service
+
+# Ideally this would be snap-specific
+/run/dockershim.sock rw,
+
+# allow managing pods' cgroups
+/sys/fs/cgroup/*/kubepods/{,**} rw,
 
-@{PROC}/sys/kernel/panic rw,
-@{PROC}/sys/kernel/panic_on_oops rw,
+# Allow tracing our own processes. Note, this allows seccomp sandbox escape on
+# kernels < 4.8
+capability sys_ptrace,
+ptrace (trace) peer=snap.@{SNAP_NAME}.*,
+
+# Allow ptracing other processes (as part of ps-style process lookups). Note,
+# the peer needs a corresponding tracedby rule. As a special case, disallow
+# ptracing unconfined.
+ptrace (trace),
+deny ptrace (trace) peer=unconfined,
+
+@{PROC}/[0-9]*/attr/ r,
+@{PROC}/[0-9]*/fdinfo/ r,
+@{PROC}/[0-9]*/map_files/ r,
+@{PROC}/[0-9]*/ns/ r,
+
+@{PROC}/sys/kernel/panic w,
+@{PROC}/sys/kernel/panic_on_oops w,
 @{PROC}/sys/kernel/keys/root_maxbytes r,
 @{PROC}/sys/kernel/keys/root_maxkeys r,
-@{PROC}/sys/vm/overcommit_memory rw,
-@{PROC}/sys/vm/panic_on_oom r,
 
-@{PROC}/diskstats r,
-@{PROC}/@{pid}/cmdline r,
+/dev/kmsg r,
+
+# kubelet calls out to systemd-run for some mounts, but not all of them and not
+# unmounts...
+capability sys_admin,
+mount /var/snap/@{SNAP_NAME}/common/{,**} -> /var/snap/@{SNAP_NAME}/common/{,**},
+mount options=(rw, rshared) -> /var/snap/@{SNAP_NAME}/common/{,**},
+
+/bin/umount ixr,
+deny /run/mount/utab rw,
+umount /var/snap/@{SNAP_INSTANCE_NAME}/common/**,
+`
+
+const kubernetesSupportConnectedPlugAppArmorKubeletSystemdRun = `
+  # kubelet mount rules
+  capability sys_admin,
+  /bin/mount ixr,
+  mount -> /var/snap/@{SNAP_INSTANCE_NAME}/common/**,
```

Comment:
@zyga - note this is a so called 'super-privileged' interface that requires an installation constraint to install and the policy can best be described as advisory (like docker-support). Also note that k8s is a general purpose container orchestration tool that is used in part to drive docker (or containerd or other backends) and the docker-support interface has a bare `mount,` rule. I observed many different mounts into the common directory and feel the above rule is as fine-grained as we want to get as I believe it will allow k8s to function, and going finer offers no security benefit since the the snap will have access to the docker (or other container runtime) socket and can simply ask it to make arbitrary mounts.

Reply:
That said, if you feel strongly about this point, I can investigate further.

Reply:
Ok, I looked at this and in very simple tests, it seems like we can use this rule:
```
mount tmpfs -> /var/snap/@{SNAP_INSTANCE_NAME}/common/**,
```

The directories in common are all random strings so the dst can't be mediated further. I'm not confident this rule won't need to be expanded in the future, but we can do that in a future iteration. I'm updating the rule now.

Reply:
This is Done.